docs/design-documents/features/storage/SecureStore/SecureStore_design.md

```diff
@@ -0,0 +1,430 @@
+# SecureStore in Mbed OS
+
+- [SecureStore in Mbed OS](#securestore-in-mbed-os)
+    + [Revision history](#revision-history)
+- [Introduction](#introduction)
+    + [Overview and background](#overview-and-background)
+    + [Requirements and assumptions](#requirements-and-assumptions)
+- [System architecture and high-level design](#system-architecture-and-high-level-design)
+  * [Design basics](#design-basics)
+    + [Data layout](#data-layout)
+    + [Basic implementation concepts](#basic-implementation-concepts)
+- [Detailed design](#detailed-design)
+    + [Class header](#class-header)
+    + [Important data structures](#important-data-structures)
+    + [Initialization and reset](#initialization-and-reset)
+    + [Core APIs](#core-apis)
+    + [Incremental set APIs](#incremental-set-apis)
+    + [Key iterator APIs](#key-iterator-apis)
+- [Usage scenarios and examples](#usage-scenarios-and-examples)
+    + [Standard usage of the class](#standard-usage-of-the-class)
+- [Other information](#other-information)
+    + [Open issues](#open-issues)
+
+
+### Revision history
+
+| Revision 	| Date           	| Authors                                                	| Mbed OS version 	| Comments         	|
+|----------	|----------------	|--------------------------------------------------------	|-----------------	|------------------	|
+| 1.0      	| 02 October 2018	| David Saada ([@davidsaada](https://github.com/davidsaada/)) 	| 5.11+           	| Initial revision 	|
+
+# Introduction
+
+### Overview and background
+
+SecureStore is a [KVStore](../KVStore/KVStore_design.md) based storage solution, providing security features on the stored data, such as encryption, authentication, rollback protection and write once, over an underlying KVStore class.
+
+### Requirements and assumptions
+
+SecureStore assumes that the underlying KVStore instance is instantiated and initialized. 
+
+# System architecture and high-level design
+
+## Design basics
+
+SecureStore is a storage class, derived from KVStore. It adds security features to the underlying key value store.
+ 
+![SecureStore Layers](./SecureStore_layers.jpg)
+
+As such, it offers all KVStore APIs. Stored data is protected with the following security options (which can be selected using the creation flags at set):  
+
+- Encryption: Data is encrypted using the AES-CTR encryption method. Nonce is randomly generated and each encryption chunk (16 bytes) uses the incrementing counter.   
+- Authentication: A 16-byte CMAC is calculated on all stored data (including metadata) and stored at the end of the record. When reading the record, calculated CMAC is compared with the stored one. In case of encryption, CMAC is calculated on the encrypted data.
```

Comment:
Key is derived from devicekey (https://github.com/ARMmbed/mbed-os/tree/master/features/device_key) using the unique ID+"AUTH" as salt.